src/coreComponents/mesh/CellElementRegionSelector.cpp

```diff
@@ -0,0 +1,265 @@
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+#include "mesh/CellElementRegionSelector.hpp"
+
+#include <fnmatch.h>
+#include <optional>
+
+
+namespace geos
+{
+using namespace dataRepository;
+using ViewKeys = CellElementRegion::viewKeyStruct;
+
+
+std::optional< string > getCellBlockAttributeValue( string_view cellBlockName )
+{
+  static constexpr string_view separator = "_";
+  return cellBlockName.find( separator ) != string_view::npos ?
+         stringutilities::removeStringAndFollowingContent( cellBlockName, separator ) :
+         std::optional< string >( std::nullopt );
+}
+
+
+CellElementRegionSelector::CellElementRegionSelector( Group const & cellBlocks ):
+  m_cellBlocks( cellBlocks )
+{
+  cellBlocks.forSubGroups< CellBlockABC >( [&] ( CellBlockABC const & cellBlock )
+  {
+    string const name = cellBlock.getName();
+    m_cellBlocksOwners.emplace( name, std::vector< CellElementRegion const * >() );
+
+    std::optional< string > regionAttributeValue = getCellBlockAttributeValue( name );
+    if( regionAttributeValue )
+    {
+      m_regionAttributeOwners.emplace( *regionAttributeValue, std::vector< CellElementRegion const * >() );
+    }
+  } );
+}
+
+
+template< typename Map, typename S = char >
+string joinMapKeys( Map map, S const & delim = S() )
+{
+  std::vector< string_view > keys;
+  keys.reserve( map.size() );
+  std::transform( map.begin(), map.end(), back_inserter( keys ),
+                  []( auto const & pair ) { return pair.first; } );
+  return stringutilities::join( keys, delim );
+}
+
+
+std::set< string >
+CellElementRegionSelector::buildMatchPatterns( CellElementRegion const & region,
+                                               std::set< integer > const & requestedAttributeValues,
+                                               std::set< string > const & requestedMatchPatterns ) const
+{
+  std::set< string > matchPatterns;
+
+  // user region attribute matching patterns creation
+  for( integer attributeValue : requestedAttributeValues )
+  {
+    string attributeValueStr = std::to_string( attributeValue );
+    // if attribute value does not exist in the mesh
+    GEOS_THROW_IF( m_regionAttributeOwners.count( attributeValueStr ) == 0,
+                   GEOS_FMT( "{}: Region attribute value '{}' not found.\nAvailable region attribute list: {{ {} }}",
+                             region.getWrapperDataContext( ViewKeys::cellBlockAttributeValuesString() ),
+                             attributeValueStr,
+                             joinMapKeys( m_regionAttributeOwners, ", " ) ),
```

Comment:
```suggestion
                             stringutilities::join( mapKeys< std::set >( map ), ", " ) ),
```

Reply:
I propose adding a `stringutilities::joinLambda( iteratorReadingLambda, list, delimiter )` function because I oftenly find myself creating new local functions to create vectors/sets to feed the `stringutilities::join()` function. I would prefer to give a lambda to give the way to obtain the values from the iterator.

Reply:
I'm not too much into it. There's a `join` version that takes iterators. We should find a way to provide the appropriate iterator instead. Maybe that will be easier with `ranges`? Meanwhile I would keep the extraction logic out of `join`?

Reply:
I think something like C++20 `std::views::keys` would be perfect here, but what about the iteration of name / `DataContext` of group arrays?
I may be wrong, but to me, giving a functor is near to give iterators, the goal is to give a way to get the data. If we give a functor to `joinLambda`, is the extraction logic really coupled?

Reply:
@TotoGaz

Reply:
You have a stream of `data`, then you can `transform` it (using `std::views::keys` or any other process for `DataContext` I imagine), then you consume it using the `join`. Your approach would merge the transformation and the consumption.
Or maybe I misunderstand your idea. Do you want to provide a functor that will return a new `string` every time you call it? How do you see the things? I'd say it's not very different from providing an iterator? Could you write pseudo code of what you want to do?

Reply:
@TotoGaz Thanks for your answers.

I've implemented it so you can see here:
`src/coreComponents/codingUtilities/StringUtilities.hpp:116`
```c++
/**
 * @brief Join strings or other printable objects returned by a lambda.
 * @tparam CONTAINER  type of container to join
 * @tparam S          type of delimiter, usually char, char const * or string
 * @tparam Lambda     type of lambda function, usually `[]( auto it ) -> string`
 * @param valueLambda  lambda function to get each values from the provided container iterator
 * @param container    container to join
 * @param delim        delimiter used to glue together strings
 * @return a string containing input values concatenated with a delimiter
 */
template< typename CONTAINER, typename S = char, typename Lambda >
string joinLamda( Lambda valueLambda, CONTAINER const & cont, S const & delim = S() )
{
```

It allows this kind of things:
`src/coreComponents/mesh/CellElementRegionSelector.cpp:124`
```c++
                   GEOS_FMT( "{}: No cellBlock named '{}'.\nAvailable cellBlock list: {{ {} }}",
                             region.getWrapperDataContext( ViewKeys::sourceCellBlockNamesString() ),
                             requestedCellBlockName,
                             stringutilities::joinLamda( []( auto pair ) { return pair->first; },
                                                         m_cellBlocksOwners, ", " ) ),
```

As `join()` can take a container or iterator to access the data, `joinLambda()` would take a functor that would give access to the data. I don't know if this is compliant with best practices?

About `std::views`, I'd like to use them but `C++20` or the `range-v3` library is needed.